### ISSUE TITLE
balance-thresholds settings do not always work correctly.

### ISSUE THREAD

**Issue by DerEwige** [ACINQ/eclair#2835]
If the balance of a channel is above the `max-htlc-value-in-flight` value of that channel.
Then the channel `max-htlc` will be set to `max-htlc-value-in-flight` even if there is a balance-thresholds rules that would set it lower.

I know that `max-htlc-value-in-flight` is an upper limit to `max-htlc` and it can never be set above this.

But I think the rules are simply applied in the wrong order.

You should first apply the `balance-thresholds` rules and only if ater those rules, the `max-htlc` is over `max-htlc-value-in-flight`, reduce it to that value.

Example:
Channel size: 1'000'000 sat
`max-htlc-value-in-flight`: 990'000 sat
Threshold rule: if below 1'024'000 sat -> Set `max-htlc` to 512'000
Sats on that channel: 1'000'000

Expected result: `max-htlc` = 512'000 sat
Observed result: `max-htlc` = 990'000 sat

eclair version: version=0.10.0 commit=a63d2c2

**Comment by t-bast** [comment]
@thomash-acinq can you take a look at this issue?

**Comment by thomash-acinq** [comment]
I've tried to reproduce it but I get `max-htlc = 512000 sat` as expected.
Is your channel private? It doesn't make sense to change `max-htlc` for private channels as the only people who can use the channel also know the exact balance.
Do you have other balance-thresholds rules? The first rule that matches is applied.

**Comment by thomash-acinq** [comment]
> ```
> {
>   available-sat = 1024000
>   max-htlc-sat =  2048000
> }
> ```

In our case this is the rule that applies. It tries to set `max-htlc` to 2048000, however that's higher than `max-htlc-value-in-flight` which is 990000, so we end up with `max-htlc = 990000`

**Comment by DerEwige** [comment]
> In our case this is the rule that applies. It tries to set `max-htlc` to 2048000, however that's higher than `max-htlc-value-in-flight` which is 990000, so we end up with `max-htlc = 990000`

that is impossible, the channel only has 1m sats it can never have over 1'024'000 sats so it should never reach that rule
this shoudl always be the last rule:

```
{
  available-sat = 1024000
  max-htlc-sat =  2048000
}
```

**Comment by DerEwige** [comment]
my channel settings:

```
	htlc-minimum-msat = 1
	max-htlc-value-in-flight-msat = 100000000000 // 1 BTC
	max-htlc-value-in-flight-percent = 99 // 80% of the channel capacity
        max-accepted-htlcs = 10
```

**Comment by thomash-acinq** [comment]
> that is impossible, the channel only has 1m sats it can never have over 1'024'000 sats so it should never reach that rule

The rule applies if the balance is below `available-sat`, not above.

**Comment by DerEwige** [comment]
> > that is impossible, the channel only has 1m sats it can never have over 1'024'000 sats so it should never reach that rule
> 
> The rule applies if the balance is below `available-sat`, not above.

yes, but the rules are applied from top to buttom right? and the first rule that hits gets applied right?
That is why they have to be in ascending order. That is how it is written in the documentation:

![image](https://github.com/ACINQ/eclair/assets/6146383/de0ab309-1aa3-4c64-8722-c5c6a2969285)

The channel has  1m sat in capacity, so the available sats on that channel will always be at maximum 1'000'000 sat right?

So if the channel has always less than 1m sat this rule shoudl always be triggered if reached.
```
{
  available-sat = 512000
  max-htlc-sat =  1024000
}
```

it should be impossible to go past this rule to the next rule, because this rule is always true for chaannels smaler than 1024000

**Comment by thomash-acinq** [comment]
No, this rule only applies if the balance is less than 512000, there is still 488000 sats to go before reaching the million.

**Comment by DerEwige** [comment]
> No, this rule only applies if the balance is less than 512000, there is still 488000 sats to go before reaching the million.

I'm sorry, you are right my rulse are wrong.
I've somehow mixed up `available-sat` and `max-htlc-sat`

I'll correct the settings right now and restart my node.
I'm sure it will work afterwards.

Sorry, this was entirely my fault

**Comment by DerEwige** [comment]
those are my new settings and they now work as expected:

```
  channel-update {
      // Balance thresholds at which to update the maximum HTLC amount
      // Must be in increasing order.
      // Set balance-thresholds = [] to disable this feature.
      balance-thresholds = [{
        available-sat = 2000  // If our balance goes below this,
        max-htlc-sat =  1000     // set the maximum HTLC amount to this (or htlc-minimum-msat if it's higher).
      },{
        available-sat = 4000
        max-htlc-sat =  2000
      },{
        available-sat = 8000
        max-htlc-sat =  4000
      },{
        available-sat = 16000
        max-htlc-sat =  8000
      },{
        available-sat = 32000
        max-htlc-sat =  16000
      },{
        available-sat = 64000
        max-htlc-sat =  32000
      },{
        available-sat = 128000
        max-htlc-sat =  164000
      },{
        available-sat = 256000
        max-htlc-sat =  128000
      },{
        available-sat = 512000
        max-htlc-sat =  256000
      },{
        available-sat = 1024000
        max-htlc-sat =  512000
      },{
        available-sat = 2048000
        max-htlc-sat =  1024000
      },{
        available-sat = 4096000
        max-htlc-sat =  2048000
      },{
        available-sat = 8192000
        max-htlc-sat =  4096000
      },{
        available-sat = 16384000
        max-htlc-sat =  8192000
      },{
        available-sat = 32768000
        max-htlc-sat =  16384000
      }]

      min-time-between-updates = 12 minutes // minimum time between channel updates because the balance changed
    }
```

**Comment by DerEwige** [comment]
user error, working as intended
